### PR TITLE
Proposed feature: Daily rewrite / Daily-saver feature

### PR DESCRIPTION
**Pull request by Xernium**
Drop a magical watch every 50 days which will automatically be used in case you overstep your streak. Not to be traded sold or bought. 

Please review this PR when done.

### CODE REVIEW

**Comment by Xernium** [comment]
I'll refactor the entire daily command next. This is more of a POC for now.

**Comment by Kodehawa** [comment]
Considering this drops every 50x streak, it seems like a fairly balanced feature. Code-wise, I'd need Natan's review here.

Also @MrLar words on this feature?

**Comment by MrLar** [comment]
lgtm feature wise

**Comment by Xernium** [comment]
Go easy on me. I need to untie the entire daily command method so it becomes a bit more maintainable

File: src/main/java/net/kodehawa/mantarobot/commands/currency/item/Items.java [hunk 1]
```diff
@@ -188,6 +188,7 @@
             // ---------------------------------- 5.4 PET ITEMS START HERE ----------------------------------
             INCUBATOR_EGG = new Item(ItemType.PET, "\uD83E\uDD5A", "Incubator Egg", "items.incubator_egg", "items.description.incubator_egg", 300, false, false, "4;3;1", 11, 12, 18),
             WATER_BOTTLE = new Item(ItemType.PET, "", "Water Bottle", "items.water_bottle", "items.description.water_bottle", 100, true, true),
+            MAGIC_WATCH = new Item(ItemType.COLLECTABLE, "⏳", "Magical Watch", "items.magic_watch", "items.description.magic_watch",0, false, false)
```
Comment (MrLar) [inline src/main/java/net/kodehawa/mantarobot/commands/currency/item/Items.java]: shouldn't this be the unicode representation of the emote for consitency sake?

Comment (Xernium) [reply src/main/java/net/kodehawa/mantarobot/commands/currency/item/Items.java]: My bad. I use eclipse instead of IntelliJ

Comment (Kodehawa) [reply src/main/java/net/kodehawa/mantarobot/commands/currency/item/Items.java]: That shouldn't matter though, I usually just look up the unicode (java) representation of it on google or something :P

**Comment by Kodehawa** [comment]
> Two questions that arise from me for this:
> 
>     * Should formatting be done using the language file?
> 
>     * Do other language files use English as fallback if they're invalid in a way? (I didn't check this one)

1. Yes, then sendMessageFormat(languageContext(key), ...)
2. Yes

**Comment by Xernium** [comment]
@Kodehawa okay got it- how should I revise this then? If I go for checking each message again it will get as messy as before

**Comment by Kodehawa** [comment]
I'll leave that up to you

**Comment by Xernium** [comment]
Out of WIP phase I guess? Idk what comes next

File: src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java [hunk 2]
```diff
@@ -108,151 +110,159 @@ public void call(GuildMessageReceivedEvent event, I18nContext languageContext, S
...
+                }
 
-                PlayerData playerData = player.getData();
-                String streak;
-                String crate = "";
+                UnifiedPlayer toAddMoneyTo = UnifiedPlayer.of(author, getConfig().getCurrentSeason());
+                User otherUser = null;
 
-                String playerId = player.getUserId();
+                boolean targetOther = mentionedUsers != null && !mentionedUsers.isEmpty();
```
Comment (natanbc) [inline src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java]: No need for the null check

File: src/main/java/net/kodehawa/mantarobot/data/Config.java [hunk 3]
```diff
@@ -64,6 +64,7 @@
     public String ipv6Block = "";
     public String excludeAddress = "";
     public int bucketFactor = 4;
+    public long daily_maxPeriod_millis = TimeUnit.HOURS.toMillis(50);
```
Comment (natanbc) [inline src/main/java/net/kodehawa/mantarobot/data/Config.java]: Fields should be `lowerCamelCase`

File: src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java [hunk 4]
```diff
@@ -108,151 +110,159 @@ public void call(GuildMessageReceivedEvent event, I18nContext languageContext, S
...
+                // Determine who gets the money
+                long dailyMoney = 150L;
+                List<User> mentionedUsers = event.getMessage().getMentionedUsers();
 
-                if (player.isLocked()) {
-                    channel.sendMessage(EmoteReference.ERROR + (mentionedUser != null ? languageContext.withRoot("commands", "daily.errors.receipt_locked") : languageContext.withRoot("commands", "daily.errors.own_locked"))).queue();
-                    return;
-                }
+                User author = event.getAuthor();
+                Player authorPlayer = MantaroData.db().getPlayer(event.getAuthor());
```
Comment (Kodehawa) [inline src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java]: You could use a field for MantaroData.db() if you'd like, easier to see.

File: src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java [hunk 5]
```diff
@@ -108,151 +110,159 @@ public void call(GuildMessageReceivedEvent event, I18nContext languageContext, S
...
 
-                    if (authorPlayerData.getDailyStreak() > 100) {
-                        authorPlayerData.addBadgeIfAbsent(Badge.BIG_CLAIMER);
+                // Not expired?
+                if(currentDailyOffset + amountStreaksavers*DAILY_VALID_PERIOD_MILLIS >= 0 ){ ;
+                    streak++;
+                    returnMessage.add(String.format(languageContext.withRoot("commands", targetOther ? "daily.streak.given.up" : "daily.streak.up" ), streak));
+                    if(currentDailyOffset<0){
+                        int streaksaversUsed = 1;
+                        for(; currentDailyOffset + streaksaversUsed*DAILY_VALID_PERIOD_MILLIS < 0 && streaksaversUsed<=streaksaversUsed;streaksaversUsed++);
```
Comment (Kodehawa) [inline src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java]: Can you clean up this? lol

Also spacing ;w;

File: src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java [hunk 6]
```diff
@@ -108,151 +110,159 @@ public void call(GuildMessageReceivedEvent event, I18nContext languageContext, S
...
+                        authorPlayer.getInventory().process(new ItemStack(Items.MAGIC_WATCH,streaksaversUsed * -1));
+                        returnMessage.add(String.format(languageContext.withRoot("commands", "daily.streak.watch_used"), streaksaversUsed, streaksaversUsed+1, amountStreaksavers-streaksaversUsed));
```
Comment (Kodehawa) [inline src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java]: spacing again, should look like `(new ItemStack(Items.MAGIC_WATCH, streakSaversUsed * -1)`

File: src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java [hunk 7]
```diff
@@ -108,151 +110,159 @@ public void call(GuildMessageReceivedEvent event, I18nContext languageContext, S
...
+                long streak = authorPlayerData.getDailyStreak();
 
-                    if (authorPlayerData.getDailyStreak() > 100) {
-                        authorPlayerData.addBadgeIfAbsent(Badge.BIG_CLAIMER);
+                // Not expired?
+                if(currentDailyOffset + amountStreaksavers*DAILY_VALID_PERIOD_MILLIS >= 0 ){ ;
+                    streak++;
+                    returnMessage.add(String.format(languageContext.withRoot("commands", targetOther ? "daily.streak.given.up" : "daily.streak.up" ), streak));
+                    if(currentDailyOffset<0){
+                        int streaksaversUsed = 1;
```
Comment (Kodehawa) [inline src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java]: Should be streakSaversUsed, Java conventions.

File: src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java [hunk 8]
```diff
@@ -108,151 +110,159 @@ public void call(GuildMessageReceivedEvent event, I18nContext languageContext, S
...
-                    }
+                long streak = authorPlayerData.getDailyStreak();
 
-                    if (authorPlayerData.getDailyStreak() > 100) {
-                        authorPlayerData.addBadgeIfAbsent(Badge.BIG_CLAIMER);
+                // Not expired?
+                if(currentDailyOffset + amountStreaksavers*DAILY_VALID_PERIOD_MILLIS >= 0 ){ ;
+                    streak++;
+                    returnMessage.add(String.format(languageContext.withRoot("commands", targetOther ? "daily.streak.given.up" : "daily.streak.up" ), streak));
+                    if(currentDailyOffset<0){
```
Comment (Kodehawa) [inline src/main/java/net/kodehawa/mantarobot/commands/MoneyCmds.java]: Spaaaaacing

**Comment by Kodehawa** [comment]
I would like a review of the new changes first, then should be ready to merge

**Comment by Kodehawa** [comment]
image

**Pull request merged**